src/io/import_off.cc

```diff
@@ -2,24 +2,234 @@
+  // XXX: are ST C N always in order?
+  boost::regex ex_cr(R"(\r$)");
+  boost::regex ex_comment(R"(\s*#.*$)");
+  boost::smatch results;
+
+  std::ifstream f(filename.c_str(), std::ios::in | std::ios::binary);
+
+  int lineno = 0;
+  std::string line;
+
+  auto AsciiError = [&](const auto& errstr){
+    LOG(message_group::Error, loc, "",
+    "OFF File line %1$s, %2$s line '%3$s' importing file '%4$s'",
+    lineno, errstr, line, filename);
+  };
+
+  auto getline_clean = [&](const auto& errstr){
+    lineno++;
+    std::getline(f, line);
+    if (f.eof()) {
+      AsciiError(errstr);
+      return false;
+    }
+    // strip DOS line endings
+    if (boost::regex_search(line, results, ex_cr))
+      line = line.erase(results.position(), results[0].length());
+    // strip comments
+    if (boost::regex_search(line, results, ex_comment))
+      line = line.erase(results.position(), results[0].length());
+    return true;
+  };
+
+  if (!f.good()) {
+    AsciiError("File error");
+    return std::make_unique<PolySet>(3);
+  }
+
+  bool got_magic = false;
+  // defaults
+  bool has_normals = false;
+  bool has_color = false;
+  bool has_textures = false;
+  bool has_ndim = false;
+  bool is_binary = false;
+  unsigned int dimension = 3;
+
+  while (line.length() == 0) {
+    if (!getline_clean("bad header: end of file"))
+      return std::make_unique<PolySet>(3);
+  }
+  if (boost::regex_search(line, results, ex_magic) > 0) {
+    got_magic = true;
+    // Remove the matched part, we might have numbers next.
+    line = line.erase(0, results[0].length());
+    has_normals = results[3].length() > 0;
+    has_color = results[2].length() > 0;
+    has_textures = results[1].length() > 0;
+    is_binary = results[6].length() > 0;
+    if (results[4].length())
+      dimension = 4;
+    if (results[5].length())
+      has_ndim = true;
+  }
+
+  // TODO: handle binary format
+  if (is_binary) {
+    AsciiError("unhandled binary format");
+    return std::make_unique<PolySet>(3);
+  }
+
+  while (line.length() == 0) {
+    if (!getline_clean("bad header: end of file"))
+      return std::make_unique<PolySet>(3);
+  }
+  std::vector<std::string> words;
+
+  if (has_ndim) {
+    boost::split(words, line, boost::is_any_of(" \t"), boost::token_compress_on);
+    if (f.eof() || words.size() < 1) {
+      AsciiError("bad header: missing Ndim");
+      return std::make_unique<PolySet>(3);
+    }
+    line = line.erase(0, words[0].length() + ((words.size() > 1) ? 1 : 0));
+    try {
+      dimension = boost::lexical_cast<unsigned int>(words[0]) + dimension - 3;
+    } catch (const boost::bad_lexical_cast& blc) {
+      AsciiError("bad header: bad data for Ndim");
+      return std::make_unique<PolySet>(3);
+    }
+  }
+
+  PRINTDB("Header flags: N:%d C:%d ST:%d Ndim:%d B:%d", has_normals % has_color % has_textures % dimension % is_binary);
+
+  if (dimension != 3) {
+    AsciiError("unhandled vertex dimensions");
```

Comment:
Could you emit the found # of dimensions too?

Reply:
OK